### PR TITLE
new: virtual tag and tag template for heatmap clicks collector feature (PPCDEV-14416)

### PR DESCRIPTION
**Pull request by rezik007**
Ticket with AC changes after last meeting and define requirements with @zawias-pro : https://piwikpro.atlassian.net/browse/PPCDEV-14357
 
This PR is based on [this one](https://github.com/PiwikPRO/solutions-templates/pull/25) however it won't be ever merged here but it will be moved to Tag's repository when approved.

Working example with virtual tag + tag template to preview in here: https://astralprojection.promilci.com/tag/app/29897930-181d-43e6-8038-42e8ca645b95/tags

@zawias-pro @musialp please remind me how we approach to browsers compatibility for that kind of tag templates? According to: https://github.com/PiwikPRO/solutions-templates/pull/25#discussion_r728736061

As always let me know if form of code (and maybe code itself) is fine for you :)

### CODE REVIEW

File: src/collectHeatmapClicksTagTemplate.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// var element = {{ Click Element }};
+  
+// element.dispatchEvent(window.PPASHeatmapClickEvent);
```
Comment (zawias-pro) [inline src/collectHeatmapClicksTagTemplate.js:3]: why is this code commented out?

File: polyfills.d.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ interface Window {
   _paq: Array<string[]>
   sevenTag: any
   heatmapEvent: Event
+  PPASHeatmapClickEvent: Event
```
Comment (zawias-pro) [inline polyfills.d.ts:5]: I'm not sure whether it's a good idea to include `PPAS` as a prefix. We already have e.g. `_ppms` that became legacy. What do you think about using `PP` instead, as the company's name is not likely to change?

File: src/collectHeatmapClicksVirtualTag.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if (tag === 'html') break;
+
+        // cut path when id exists
+        if (element.id) {
+          tag = `#${element.id}`;
+          stringSteps.push(tag);
+          break;
+        }
+
+        // if parent has multiple same tags so we should defined which child it is
```
Comment (zawias-pro) [inline src/collectHeatmapClicksVirtualTag.js:30]:
```suggestion
        // if parent has multiple same tags so we should define which child it is 
```

**Comment by zawias-pro** [comment]
> @zawias-pro @musialp please remind me how we approach to browsers compatibility for that kind of tag templates? According to: #25 (comment)

Technically we should support IE8, but we're actively working on moving browser support to IE11+. For now, your "deprecated" approach is the way to go.

File: src/collectHeatmapClicksVirtualTag.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        stringSteps.push(tag);
+      }
+
+      return stringSteps.reverse().join('>');
+    };
+    let targets = [];
+    let finalPath = '';
+
+    // handling missing method in IE11 / Edge
```
Comment (zawias-pro) [inline src/collectHeatmapClicksVirtualTag.js:57]:
```suggestion
    // handling missing method in IE11
```
Edge supports `event.composedPath`

**Review by zawias-pro — APPROVED** [review]
Everything works fine. I'm going to move it to https://github.com/PiwikPRO/PPMS-TagManager-Api/tree/develop/app/Resources/third_party_tag_templates myself. Thanks for your help @rezik007 :)

**Comment by rezik007** [comment]
Closing because of code was fixed by Tag team themselfs and moved to another repository :)

**Pull request closed without merging**